### PR TITLE
fix: prerendered root +server.js endpoints use dynamic file extensions, warn user if not html

### LINKED ISSUE #11087: Root `+server.js` that is prerendered is always saved as `index.html`
### Describe the bug

Even when `is_html` is false (e.g., content-type is not text/html), the prerendered root endpoint (src/routes/+server.js) gets saved as `index.html`. This is caused by `path.slice(config.paths.base.length + 1)` returning an empty string which is falsy, so the `'index.html` filename gets returned.

https://github.com/sveltejs/kit/blob/4559e66b649516cc996a042a21739fa165388335/packages/kit/src/core/postbuild/prerender.js#L145-L153

### Reproduction

n/a

### Logs

_No response_

### System Info

```Shell
n/a
```

### Severity

serious, but I can work around it

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by aewing**
## Summary

Fixes #11087

When a root `+server.js` endpoint is prerendered with non-HTML content, it was incorrectly saved as `index.html` due to a fallback in the `output_filename` function.

## The Problem

The issue occurs when:
1. Someone has a root `+server.js` (without a `+page.svelte`)
2. The endpoint has `export const prerender = true`
3. The endpoint returns non-HTML content (like JSON or plain text)

Previously, the code would always default to `index.html` for the root path, regardless of content type.

## The Solution

This PR:
- Only defaults to `index.html` for HTML content at the root path
- Uses content-type headers to determine appropriate file extensions for non-HTML content:
  - `application/json` → `index.json`
  - `text/plain` → `index.txt`
  - Other common types are mapped appropriately
- Adds a warning when prerendering root `+server.js` endpoints, since most static hosts expect `index.html` at the root
- Falls back to `index` (no extension) for unrecognized content types

## Tests

Added tests to verify:
- JSON endpoints are saved without `.html` extension
- Plain text endpoints are saved without `.html` extension
- The fix doesn't break existing functionality

---

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### Edits

- [x] Please ensure that 'Allow edits from maintainers' is checked. PRs without this option may be closed.

### CODE REVIEW

**Comment by aewing** [comment]
@eltigerchino thoughts on this approach?

File: packages/kit/src/core/postbuild/prerender.js [hunk 1]
```diff
@@ -360,9 +398,19 @@ async function prerender({ hash, out, manifest_path, metadata, verbose, env }) {
...
 
-		const file = output_filename(decoded, is_html);
+		const file = output_filename(decoded, is_html, type);
 		const dest = `${config.outDir}/output/prerendered/${category}/${file}`;
 
+		// Warn about prerendering non-HTML root endpoints
+		if (decoded === config.paths.base + '/' && !is_html && category === 'pages') {
+			log.warn(
+				`Prerendering root +server.js endpoint with content-type '${type}'. ` +
+					`Note: Most static hosts expect 'index.html' at the root. ` +
```
Comment (teemingc) [inline packages/kit/src/core/postbuild/prerender.js]: Not sure if we need this warning because the user might not be deploying to a static host. e.g., I can configure to prerender the root endpoint but I don't have to prerender all other pages (unless the user is using the static adapter which is not guaranteed here).

File: .changeset/puny-doodles-taste.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@sveltejs/kit': patch
+---
+
+Endpoints pre-rendered at root (routes/+server.{js|ts}) generate index.{content-type extension}, i.e. index.json, index.txt
```
Comment (teemingc) [inline .changeset/puny-doodles-taste.md]:
```suggestion
fix: endpoints prerendered at the root path (`routes/+server.{js|ts}`) use the correct file extension in the filename based on the content-type header returned (i.e. index.json for application/json)
```

File: packages/kit/src/core/postbuild/prerender.js [hunk 3]
```diff
@@ -176,6 +190,31 @@ async function prerender({ hash, out, manifest_path, metadata, verbose, env }) {
...
+			'text/css': '.css',
+			'text/javascript': '.js',
+			'application/javascript': '.js',
+			'application/xml': '.xml',
+			'text/xml': '.xml',
+			'application/pdf': '.pdf',
+			'image/svg+xml': '.svg'
+		};
+
+		return mime_to_ext[mime] || '';
```
Comment (teemingc) [inline packages/kit/src/core/postbuild/prerender.js]: We can probably use the `mrmime` package that's already installed to do this. So something like:
```js
import { mimes } from 'mrmime';

// TODO: `mimes` is a map of file extensions to mime types but we want it the other way around
```

File: packages/kit/src/core/postbuild/prerender.js [hunk 4]
```diff
@@ -165,9 +165,23 @@ async function prerender({ hash, out, manifest_path, metadata, verbose, env }) {
...
+	 * @param {string} [content_type]
 	 */
-	function output_filename(path, is_html) {
-		const file = path.slice(config.paths.base.length + 1) || 'index.html';
+	function output_filename(path, is_html, content_type) {
+		const file = path.slice(config.paths.base.length + 1);
+
+		// Special handling for root path
+		if (!file) {
+			if (is_html) {
```
Comment (teemingc) [inline packages/kit/src/core/postbuild/prerender.js:176]: Looking at how `is_html` is evaluated, I'm wondering why a redirect response automatically considers the file as a html file. We might need to rethink that one.

File: packages/kit/src/core/postbuild/prerender.js [hunk 5]
```diff
@@ -165,9 +165,23 @@ async function prerender({ hash, out, manifest_path, metadata, verbose, env }) {
...
+				return 'index.html';
+			}
+			// For non-HTML root endpoints, determine extension from content-type
+			if (content_type) {
+				const ext = get_extension_from_content_type(content_type);
+				return ext ? `index${ext}` : 'index';
+			}
```
Comment (teemingc) [inline packages/kit/src/core/postbuild/prerender.js:183]: I wonder if we need this logic for non-root paths as well. We're already appending `.html` to paths missing it if the content-type header is html.

File: packages/kit/test/prerendering/root-server-endpoint/test/tests.spec.js
Comment (teemingc) [inline packages/kit/test/prerendering/root-server-endpoint/test/tests.spec.js:1]: We probably only need one test app for this and it could be used for edge cases like this one where there is no root page. (So we might rename this to prerendering/no-root-page , etc.)

File: packages/kit/src/core/postbuild/prerender.js (hunk 3 above)
Comment (aewing) [reply packages/kit/src/core/postbuild/prerender.js]: I actually looked into this, but from what I recall mrmime doesn't export mime -> ext it's ext -> mime, so then we have to dedupe in some cases (jpeg, jpg) -- I felt like this was a safer API because we could easily reason about our commitments and support this way -- but happy to reverse-map the mrmime package and go with first/last match if that's what we'd prefer.

Comment (aewing) [reply packages/kit/src/core/postbuild/prerender.js]: jpeg/jpg is a weird example but you get my meaning

**Comment by Rich-Harris** [comment]
This... doesn't seem like a good idea to me at all. The point of a root route is that you can access it by visiting `/`, but in this case you're supposed to visit `/` in development but `/index.json` (or whatever) in production? That makes no sense. What am I missing?

Combine that with all the extra complexity this PR entails, and the fact that we have to add a whole new test app (which will slow down every CI run) it's a hard 👎 from me. I _would_ accept a PR that threw an error if a prerenderable root `+server.js` responded with a non-HTML response.

**Comment by aewing** [comment]
> This... doesn't seem like a good idea to me at all. The point of a root route is that you can access it by visiting `/`, but in this case you're supposed to visit `/` in development but `/index.json` (or whatever) in production? That makes no sense. What am I missing?
> 
> Combine that with all the extra complexity this PR entails, and the fact that we have to add a whole new test app (which will slow down every CI run) it's a hard 👎 from me. I _would_ accept a PR that threw an error if a prerenderable root `+server.js` responded with a non-HTML response.

Yea, this one felt a bit off to me. 
I will leave this open in case anyone wants to make an argument in favor of this solution, but I will open an alternative where I throw an error instead of the warning I had initially and leave it at that.

**Comment by teemingc** [comment]
> This... doesn't seem like a good idea to me at all. The point of a root route is that you can access it by visiting `/`, but in this case you're supposed to visit `/` in development but `/index.json` (or whatever) in production? That makes no sense. What am I missing?
> 
> Combine that with all the extra complexity this PR entails, and the fact that we have to add a whole new test app (which will slow down every CI run) it's a hard 👎 from me. I _would_ accept a PR that threw an error if a prerenderable root `+server.js` responded with a non-HTML response.

I'm fine with that. It doesn't seem like this was a problem anyone ran into since the issue had very few responses. I also can't remember where that original discussion was that made me file the issue in the first place.
